llext: Support non-paired RISC-V PCREL Relocation

WorldofJARcraft opened the pull request:
Currently, RISC-V's architecture-specific relocations assume that all relocations of type `R_RISCV_PCREL_LO12_I` and `_S` are processed immediately after the `R_RISCV_PCREL_HI20` relocation that they share a relocation target with. While this is the case most of the time, the RISC-V PSABI specification does not guarantee that. This commit corrects this by determining the `R_RISCV_PCREL_HI20` relocation based on the symbol value of the `R_RISCV_PCREL_LO12` relocation, as specified in the [PSABI](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#pc-relative-symbol-addresses).

WorldofJARcraft committed:
llext: Refactor symbol lookup logic into function

This commit refactors logic in llext_link.c that resolves
a symbol to be imported by an llext into a utility function.
Thereby, the logic can be re-used in other functions if
needed, e.g., when resolving symbols for indirect relocations.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+		 * For regular symbols, the link address is obtained by
+		 * adding st_value to the start address of the section
+		 * in which the target symbol resides.
+		 */
```
You're saving 1 level of indentation here, so these lines could be reformatted for a bit easier reading

WorldofJARcraft replied:
Reformatted

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -393,64 +595,25 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, const struct llext_l
-					return -ENODATA;
-				}
+			ret = llext_lookup_symbol(&link_addr, &rel, &sym, ext, ldr, name, shdr, i,
+						  j);
```
would it fit within 100 characters if you join the two lines? You need to check the error code

WorldofJARcraft replied:
Added the error check and reformatted the line

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+	} else {
+		LOG_ERR("rela section %d, entry %d: cannot apply relocation: "
+			"target symbol has unexpected section index %d (0x%X)",
+			section_idx, entry_idx, sym->st_shndx, sym->st_shndx);
```
`section_idx` and `entry_idx` are only used in this print. Move it out of the function and remove the function arguments.

WorldofJARcraft replied:
Done

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+			struct llext *dep;
+
+			*link_addr = (uintptr_t)llext_find_extension_sym(name, &dep);
+			if (link_addr) {
```
missing dereference asterisk `*`

WorldofJARcraft replied:
Added *

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+			       struct llext *ext, struct llext_loader *ldr, const char *name,
+			       const elf_shdr_t *shdr, int section_idx, int entry_idx)
+{
+	*link_addr = 0;
```
this isn't in the original code and is superfluous - as long as you add an error check in the caller

WorldofJARcraft replied:
Removed

lyakh reviewed (dismissed):
Some clean up around function return codes and arguments is needed, some further comments are optional

WorldofJARcraft commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+	*link_addr = 0;
+
+	if (ELF_R_SYM(rel->r_info) == 0) {
+		/* no symbol ex: R_ARM_V4BX relocation, R_ARM_RELATIVE  */
```
Updated the comment

WorldofJARcraft commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+ */
+static int llext_lookup_symbol(uintptr_t *link_addr, const elf_rela_t *rel, const elf_sym_t *sym,
+			       struct llext *ext, struct llext_loader *ldr, const char *name,
+			       const elf_shdr_t *shdr, int section_idx, int entry_idx)
```
Changed it accordingly in both new functions

pillo79 reviewed (commented):
Agree with all comments from @lyakh. 
Also, for easier review, can you split the refactor that creates  `llext_lookup_symbol` in a separate commit? Thanks!

WorldofJARcraft commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +143,206 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+					     elf_word rel_cnt, uintptr_t sect_base, int section_idx,
+					     int entry_idx, struct llext *ext, const elf_sym_t *sym,
+					     struct llext_loader *ldr, uintptr_t *link_addr_out,
+					     int *last_rel_idx)
```
I have removed the parameters that can be computed efficiently.

WorldofJARcraft commented:
> Agree with all comments from @lyakh. Also, for easier review, can you split the refactor that creates `llext_lookup_symbol` in a separate commit? Thanks!

I have split the refactor into a separate commit.

pillo79 reviewed (commented):
Thanks for the refactor! 
This PR adds a lot of RISCV-specific stuff to the common LLEXT code; I think it's a pity and I'm not yet convinced this is necessary.

If the `llext_lookup_symbol()` function was exported in `include/llext/llext_internal.h`, it should be possible to move the whole `llext_riscv_find_sym_pcrel()` function to the arch-specific `arch/riscv/core/elf.c` and only invoke it when those relocations are encountered. 
But I am probably missing something big! :slightly_smiling_face:

WorldofJARcraft commented:
I can move the new `llext_riscv_find_sym_pcrel` function into the architecture-specific `elf.c`, that would not be a problem. You are right, the function is architecture-specific.
However, the function needs access to the loader and the extension, which are currently not provided via the `arch_elf_relocate` api, which is why it needs to be called from `llext_link`. I can add these additional parameters to `arch_elf_relocate` and remove the RISCV-specific code from `llext_link.c` if you prefer.

WorldofJARcraft commented:
I believe the RISC-V PSABI breaks one of the fundamental assumptions that were made when originally designing the API: the `PCREL_LO12` relocations cannot be relocated with just the information in the ELF relocation entry and symbol. The symbol value for these relocations is actually an instruction offset from the beginning of the current segment. The instruction at this offset is of type `PCREL_HI20`. The immediate for the `PCREL_LO12` instruction is the lower 12 bits of the relocation target computed for the `PCREL_HI20` relocation; together, these two types of relocation allow accesses within a 32-bit distance from the `PCREL_HI20` relocation. 
This causes the following problem: when a `PCREL_LO12` relocation is encountered, I need to find the `PCREL_HI20` relocation and compute the relocation target from here, using the lower 12 bits for the `PCREL_LO12` relocation. To be able to do that, I need the additional parameters that `arch_elf_relocate` does not currently provide.

WorldofJARcraft commented:
My original code used the assumption that the `PCREL_LO12` is always processed immediately after the matching `PCREL_HI20`. This allowed me to store the jump target in a variable.
However, I realized this is not guaranteed by the specification. I also managed to break this with one of my llext extensions.

pillo79 commented:
> I believe the RISC-V PSABI breaks one of the fundamental assumptions that were made when originally designing the API: the `PCREL_LO12` relocations cannot be relocated with just the information in the ELF relocation entry and symbol.

Yep, never saw something like this before. And AFAICS it's even worse: according to your ABI link, surprisingly not even "HI20 is always earlier than its LO12" is guaranteed, so you couldn't even assume that you have already computed HI20 and can reload those bits from the target. Very unlucky situation.

> I can add these additional parameters to `arch_elf_relocate` and remove the RISCV-specific code from `llext_link.c` if you prefer.

The API was designed to hide that information at such a low level stage, but this looks like a valid exception. Personally I would add these (also, they actually are already there for Xtensa...), but I'd like to hear other opinions on this as well.

WorldofJARcraft commented:
I am pretty sure that the instruction with the `HI20` relocation will always precede the `LO12` instruction *in code*. The instructions that generate `HI20`, `auipc` and `lui`, always overwrite a register, while `LO12` instructions like `lw`  or `add` have a register input. I believe this order is also intended in the specification:

> The successive instruction has a signed 12-bit immediate so the value of the preceding high 20-bit relocation may have 1 added to it.

However, the compiler can generate the relocation entries in any order (this is how I encountered the issue in the first place), and there might be other instructions with all possible relocations including `HI20`  and `LO12` between a `HI20`  and `LO12` pair in code.

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -479,6 +595,26 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, const struct llext_l
+				}
+			} else {
+				/* unused */
+				(void)last_rel_idx;
```
does the compiler actually need this? Are there warnings without it?

WorldofJARcraft replied:
GCC would definitely complain if this was an ifdef rather than an if.

pillo79 reviewed (dismissed):
@teburd @lyakh pausing merge to get feedback to my open question (https://github.com/zephyrproject-rtos/zephyr/pull/82799#pullrequestreview-2495397019 and below discussion)

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -219,6 +220,120 @@ static int llext_lookup_symbol(struct llext *ext, struct llext_loader *ldr, uint
+ * @param last_rel_idx index of last relocation entry processed in search
+ *
+ */
+static inline int llext_riscv_find_sym_pcrel(struct llext *ext, struct llext_loader *ldr,
```
...and let's remove `inline` - I don't see too many of them used under subsys/llext, the compiler will figure out by itself

WorldofJARcraft replied:
Done

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -219,6 +220,120 @@ static int llext_lookup_symbol(struct llext *ext, struct llext_loader *ldr, uint
+		 * is actually the offset of the PCREL_HI20 relocation instruction from section
+		 * start
+		 */
+		if (!(candidate.r_offset == sym->st_value && reloc_type == R_RISCV_PCREL_HI20)) {
```
I think this would be more readable as
```
	if (candidate.r_offset == sym->st_value && reloc_type == R_RISCV_PCREL_HI20) {
		break;
	}
	if (i + 1 >= rel_cnt) {
		/* wrap around and search in previously processed indices as well */
		i = -1;
	}
}
/* we found a match - need to compute the relocation for this instruction */
/* lower 12 bits go to the PCREL_LO12 relocation */
...
```
and if you modify the loop as proposed above, then you don't need the wrap-around part either.

lyakh commented:
> @teburd @lyakh pausing merge to get feedback to my open question ([#82799 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/82799#pullrequestreview-2495397019) and below discussion)

@pillo79 yeah, would be cleaner

WorldofJARcraft commented:
In order to make reviewing easier, I will make two force-pushes:
- the first one incorporates the changes requested by @lyakh 
- the second one moves the `llext_riscv_find_sym_pcrel` function to the RISC-V architecture-dependent elf.c. @pillo79 I think I misunderstood your comment; I originally thought you wanted me to integrate the logic in `llext_riscv_find_sym_pcrel` function into `arch_elf_relocate` for RISC-V. This would require extensive API changes. However, just moving the function to elf.c is easy to do and does not affect the other architectures.

teburd commented:
@pillo79 @WorldofJARcraft in reviewing this a lot more closely this morning I think this is pretty much a killer for the arch_elf_relocate API we have today which *assumes* the linker code can be generic providing all the needed parameters, clearly its not the case for RISCV

"The R_RISCV_PCREL_LO12_I or R_RISCV_PCREL_LO12_S relocations contain a label pointing to an instruction in the same section with an R_RISCV_PCREL_HI20 relocation entry that points to the target symbol"

This really means the "symbol" in the calculation for this relocation is located in a *second* relocation. This is actually fairly painful if I'm understanding this correctly.

I believe this really means we need to extend our arch API for relocations to accept more of the information from the ELF... as noted in this PR.

I agree with @pillo79 though that adding arch specifics to llext is likely the wrong place, despite my initial +1 (I'm generally all in favor of moving fast) in this case.

WorldofJARcraft commented:
Merging `llext_riscv_find_sym_pcrel` into `arch_elf_relocate` might result in an API like this:
```C
int arch_elf_relocate(struct llext *ext, struct llext_loader *ldr,
	              const elf_rela_t *rel, const elf_shdr_t *shdr,
	              const elf_sym_t *sym, uintptr_t loc, uintptr_t sym_base, 
                      const char *sym_name, uintptr_t load_bias);
```
The last four parameters can be computed using the new parameters, but passing them directly like this when possible reduces read operations, so it makes sense to keep them.
This is why I was hesitant to implement this.

teburd commented:
> Merging `llext_riscv_find_sym_pcrel` into `arch_elf_relocate` might result in an API like this:
> 
> ```c
> int arch_elf_relocate(struct llext *ext, struct llext_loader *ldr,
> 	              const elf_rela_t *rel, const elf_shdr_t *shdr,
> 	              const elf_sym_t *sym, uintptr_t loc, uintptr_t sym_base, 
>                       const char *sym_name, uintptr_t load_bias);
> ```
> 
> The last four parameters can be computed using the new parameters, but passing them directly like this when possible reduces read operations, so it makes sense to keep them. This is why I was hesitant to implement this.

It is what it is, if the information is needed in the function all then it must be passed in. Unless someone has a proposal for something simpler, I don't see a massive issue with this. We're looking a lot closer to how any other linker looks in this regard afaik.

WorldofJARcraft committed:
llext: Add parameters to arch_elf_relocate

The RISC-V port of llext requires additional parameters for
handling non-adjacent HI20/LO12 relocations in arch_elf_relocate():
the current extension (struct llext), the current extension loader
(struct llext_loader), the current section header (elf_shdr_t) and
the current symbol (elf_sym_t).
This changes the signature of arch_elf_relocate accordingly.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

WorldofJARcraft commented:
I have changed the API accordingly in a new commit.
I have realized that the `load_bias` is trivial to compute, so I moved it to a new inline helper function.
However, as I pointed out before, I have kept the symbol name and base address parameters for efficiency.
The new `llext_lookup_symbol` function is now static in RISC-V's elf.c.

lyakh commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -11,15 +11,20 @@
+static inline const char *llext_string(const struct llext_loader *ldr, const struct llext *ext,
+	enum llext_mem mem_idx, unsigned int idx)
+{
+return (char *)ext->mem[mem_idx] + idx;
```
indentation

lyakh commented on subsys/llext/llext_link.c:
```diff
@@ -142,6 +142,84 @@ static const void *llext_find_extension_sym(const char *sym_name, struct llext *
+		 * address of the section in which the target symbol resides.
+		 */
+		*link_addr =
+			(uintptr_t)llext_loaded_sect_ptr(ldr, ext, sym->st_shndx) + sym->st_value;
```
should easily fit in a single line

WorldofJARcraft replied:
Done

lyakh reviewed (commented):
just some cosmetic clean up

WorldofJARcraft commented on subsys/llext/llext_link.c:
```diff
@@ -220,7 +208,6 @@ static int llext_lookup_symbol(struct llext *ext, struct llext_loader *ldr, uint
 	return 0;
 }
 
-
```
Done

WorldofJARcraft commented on arch/riscv/core/elf.c:
```diff
@@ -61,6 +64,134 @@ static inline int riscv_relocation_fits(long long jump_target, long long max_dis
+	symbol_name = llext_string(ldr, ext, LLEXT_MEM_STRTAB, candidate_sym.st_name);
+
+	ret = llext_lookup_symbol(ext, ldr, &link_addr, &candidate, &candidate_sym,
+	symbol_name, shdr);
```
Done

WorldofJARcraft commented on subsys/llext/llext_link.c:
```diff
@@ -494,6 +481,7 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, const struct llext_l
 				return ret;
 			}
 
+
```
Done

pillo79 commented on include/zephyr/llext/llext.h:
```diff
@@ -354,18 +354,24 @@ int llext_add_domain(struct llext *ext, struct k_mem_domain *domain);
  */
-int arch_elf_relocate(elf_rela_t *rel, uintptr_t loc,
-			     uintptr_t sym_base_addr, const char *sym_name, uintptr_t load_bias);
+int arch_elf_relocate(const struct llext *ext, struct llext_loader *ldr, elf_rela_t *rel,
```
Please swap `ldr` and `ext` so that they match every other LLEXT function call :innocent:

WorldofJARcraft replied:
Done

pillo79 commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -28,6 +33,28 @@ struct llext_elf_sect_map {
 
 const void *llext_loaded_sect_ptr(struct llext_loader *ldr, struct llext *ext, unsigned int sh_ndx);
 
+static inline uintptr_t llext_load_bias(const struct llext *ext)
+{
+	return (uintptr_t)ext->mem[LLEXT_MEM_TEXT];
+}
```
The `load_bias` _is_ the start of the text section. Since now we have direct access to the `ext` parameter in the relocation functions, I think it's clearer to write that explicitly where it is used, instead of adding this function.

WorldofJARcraft replied:
I have renamed it "llext_text_start"

pillo79 commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -28,6 +33,28 @@ struct llext_elf_sect_map {
+ *
+ * @return 0 for OK, negative for error
+ */
+int llext_lookup_symbol(struct llext *ext, struct llext_loader *ldr, uintptr_t *link_addr,
```
Swap `ldr` and `ext` here as well

pillo79 commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -28,6 +33,28 @@ struct llext_elf_sect_map {
+	return (uintptr_t)ext->mem[LLEXT_MEM_TEXT];
+}
+
+/**
+ * @brief Determine address of a symbol.
+ *
+ * @param ext llext extension
+ * @param ldr llext loader
+ * @param link_addr (output) resolved address
+ * @param rel relocation entry
+ * @param sym symbol entry
+ * @param name symbol name
+ * @param shdr section header
+ *
+ * @return 0 for OK, negative for error
+ */
```
Nit: the Doxygen tags are ignored because this is explicitly excluded from documentation build (see the `@endcond` just below). A simpler comment can have the same effect :slightly_smiling_face:

WorldofJARcraft replied:
Done

pillo79 reviewed (commented):
Thanks, indeed looks way nicer to me! :+1:
Just a few minor nits and a general remark, it looks like some indentation is off: tabs should be 8 chars in width and extra parameters should be indented all the way just after the matching opening brace.

WorldofJARcraft commented:
I will have to force-push again to fix the indentation. My editor is playing games with me.

pillo79 reviewed (dismissed):
Not a fan of the new `llext_text_start` but otherwise LGTM, thanks for your efforts!

teburd reviewed (commented):
Since we now have the symbol table accessible to us from the loader, perhaps giving an index into the symbol table is enough rather than the elf_sym_t, sym_name, and sym_base_addr?

Those things could be obtained via accessors/helper functions.

WorldofJARcraft committed:
tests: llext: Test RISC-V non-paired relocation

Tests an edge case in the RISC-V PSABI: In the medany and the medlow code
models, the compiler emits auipc/lui (U-type) and ld/sw (I-type/S-type)
instruction pairs for accessing a non-local symbol.
The U-type instruction sets the upper 20 bits, the I/S-type the lower 12.
The U-type and I-type/S-type instruction pairs are often adjacent in code.
This is also what the current llext architecture-specific relocations
expect.
However, this need not be the case - compilers can re-use the upper 20
bits set by the U-type instruction with multiple I/S-type instructions,
which is a useful optimization for multiple loads/stores of or within
the same symbol.
This commit adds a unit test for this behavior, which currently fails
for RISC-V.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

WorldofJARcraft commented:
> Since we now have the symbol table accessible to us from the loader, perhaps giving an index into the symbol table is enough rather than the elf_sym_t, sym_name, and sym_base_addr?
> 
> Those things could be obtained via accessors/helper functions.

The arguments `sym` and `link_addr` require a read, so I would like to keep them.
However, I will eliminate the `loc`, `sym_base_addr`, `sym_name` arguments and replace them with helper functions, as they can be computed from data that is already in memory.

pillo79 commented:
> [...] there appears to be an assertion failure in an unrelated test on RISC-V (`test_inspect`, probably due to linker flags).

Thanks for noticing, did not realize RISC-V was disabled in Twister! (why?) 
Submitted a fix for this in #85989.

WorldofJARcraft commented:
RISC-V tests currently do not run because of the `min_ram` and `min_flash` attributes in testcase.yaml. The qemu "boards" for RISC-V do not specify RAM and flash, so twister deems them unsuitable.

pillo79 commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -28,6 +28,49 @@ struct llext_elf_sect_map {
 
 const void *llext_loaded_sect_ptr(struct llext_loader *ldr, struct llext *ext, unsigned int sh_ndx);
 
+static inline uintptr_t llext_text_start(const struct llext *ext)
+{
+	return (uintptr_t)ext->mem[LLEXT_MEM_TEXT];
+}
```
IMHO there is no gain in adding such a specific private function; ELF code can freely access internals of the `struct llext` and doing so would make the code more similar to the rest of the LLEXT subsystem.

WorldofJARcraft replied:
Done

pillo79 reviewed (commented):
Comments on the new functions in `llext_internal.h` should be addressed IMO. 

Not blocking since the latest changes were requested by Tom, but I feel that this will bring LLEXT to a more fragmented situation and I would like to avoid that if possible.

teburd commented on subsys/llext/llext_link.c:
```diff
@@ -382,99 +467,9 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, const struct llext_l
-
-			name = llext_string(ldr, ext, LLEXT_MEM_STRTAB, sym.st_name);
-
-			LOG_DBG("relocation %d:%d info 0x%zx (type %zd, sym %zd) offset %zd "
-				"sym_name %s sym_type %d sym_bind %d sym_ndx %d",
-				i, j, (size_t)rel.r_info, (size_t)ELF_R_TYPE(rel.r_info),
-				(size_t)ELF_R_SYM(rel.r_info), (size_t)rel.r_offset,
-				name, ELF_ST_TYPE(sym.st_info),
-				ELF_ST_BIND(sym.st_info), sym.st_shndx);
```
This can still be done but would require calling the accessors (cheap). The LOG_DBG and lookups can be wrapped in an ifdef CONFIG_LLEXT_LOG_LEVEL_DBG to avoid doing stuff we don't need to when its not debug logging.

WorldofJARcraft commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -28,6 +28,49 @@ struct llext_elf_sect_map {
+	return (const char *)ext->mem[mem_idx] + idx;
+}
+
+static inline uintptr_t llext_get_section_base(struct llext_loader *ldr, struct llext *ext,
+					       const elf_shdr_t *shdr)
```
Done

WorldofJARcraft commented on include/zephyr/llext/llext_internal.h:
```diff
@@ -28,6 +28,49 @@ struct llext_elf_sect_map {
+
+static inline uintptr_t llext_get_reloc_instruction_location(struct llext_loader *ldr,
+							     struct llext *ext,
+							     const elf_shdr_t *shdr,
```
Done

teburd reviewed (dismissed):
Patch set looks really nice now, thank you for the hard work!

WorldofJARcraft committed:
llext: Support non-paired RISC-V PCREL Relocation

Currently, RISC-V's architecture-specific relocations assume that
all relocations of type R_RISCV_PCREL_LO12_I and _S are processed
immediately after the R_RISCV_PCREL_HI20 relocation that they
share a relocation target with. While this is the case most of
the time, the RISC-V PSABI specification does not guarantee that.
This commit corrects this by determining the R_RISCV_PCREL_HI20
relocation based on the symbol value of the R_RISCV_PCREL_LO12
relocation, as specified in the PSABI.

Signed-off-by: Eric Ackermann <eric.ackermann@cispa.de>

pillo79 reviewed (dismissed):
Looks awesome. Thank you for your efforts, much appreciated! :bow:

WorldofJARcraft commented:
I added an `ifdef` guard to cover cases where `CONFIG_LOG` is disabled - this should fix the checks.
